rmw_connext_shared_cpp/src/init.cpp

```diff
@@ -12,23 +12,242 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <mutex>
+
 #include "rmw_connext_shared_cpp/init.hpp"
-#include "rmw_connext_shared_cpp/ndds_include.hpp"
 
+#include "rcutils/get_env.h"
 #include "rmw/error_handling.h"
 
+#include "rmw_connext_shared_cpp/ndds_include.hpp"
+
+#include "./qos_impl.hpp"
+
+/// Flag used to make sure \ref init() only runs once.
+static std::once_flag g_run_once_flag;
+/// Return value of \ref is_ros_qos_ignored().
+static bool g_is_ros_qos_ignored = false;
+/// Return value of \ref does_node_profile_override().
+static bool g_are_node_profiles_allowed = false;
+/// Return value of \ref is_publish_mode_overriden().
+static bool g_is_publish_mode_overriden = true;
+
+/// Tri-state retcode used in `set_default_qos_library` and `is_env_variable_set`.
+enum class TristateRetCode {SET, NOT_SET, FAILED};
+
+/// Set the default Connext qos profile library.
+/**
+ * If `RMW_CONNEXT_QOS_PROFILE_LIBRARY` environment variable is set,
+ * the library matching that name is used.
+ * If only one user provided library is found, that library is set as the default one.
+ * If not, not qos profile library is set.
+ *
+ * \return `TristateRetCode::SET` if the default library was set, or
+ * \return `TristateRetCode::NOT_SET` if the default library was not set, or
+ * \return `TristateRetCode::FAILED` if failed to read the environment variable
+ *  or to set the default library.
+ */
+static TristateRetCode
+set_default_qos_library(DDS::DomainParticipantFactory * dpf);
+
+/// Set the default Connext qos profile.
+/**
+ * If `RMW_CONNEXT_DEFAULT_QOS_PROFILE`, that profiles is set as the default.
+ * It uses the globally set default library.
+ *
+ * \return false if failed to get the environment variable or to set the default profile,
+ *  else false.
+ */
+static bool
+set_default_qos_profile(DDS::DomainParticipantFactory * dpf);
+
+/// Check if environment variable is set.
+/**
+ * Check if the provided environment variable is "1".
+ *
+ * \return `TristateRetCode::SET` if the environment variable of the provided name is equal to "1".
+ * \return `TristateRetCode::NOT_SET` if the environment variable has another value or is unset, or
+ * \return `TristateRetCode::FAILED` if failed to read the environment variable.
+ */
+static TristateRetCode
+is_env_variable_set(const char * env_var_name);
+
 rmw_ret_t
 init()
 {
-  DDS::DomainParticipantFactory * dpf_ = DDS::DomainParticipantFactory::get_instance();
-  if (!dpf_) {
-    RMW_SET_ERROR_MSG("failed to get participant factory");
-    return RMW_RET_ERROR;
+  rmw_ret_t ret = RMW_RET_OK;
```

Comment:
Should this be `static`? Otherwise, a second call to `init()` will always return `RMW_RET_OK`, even if the first call returned `RMW_RET_ERROR`.

Reply:
Great catch, fixed in https://github.com/ros2/rmw_connext/pull/460/commits/0cb33be75001cca34a31b5920cb0344bebd5e3c8.